### ISSUE TITLE
Trying to stream raw PCM data from a process into pyatv

### ISSUE THREAD

**Issue by daredoes** [postlund/pyatv#1997]
### What do you need help with?

Hello!

I am attempting to use Pyatv to read a buffer from another process as it comes in. Currently, I am trying to load the process using asyncio subprocess with something like

```python
async def create_process(cmd, *args):
    process = await asp.create_subprocess_exec(
        cmd, stdin=None, stdout=asp.PIPE, stderr=None, *args)
    return process
```

Then later I try to read in the data from `stdout` and pass that to the pyatv stream file function

```python
process = await create_process("./snapclient.sh", f"--player file --logsink null")
await asyncio.sleep(3)
try:
    print("* Starting to stream")
    while process:
      await atv.stream.stream_file(process.stdout)
      await asyncio.sleep(0.01)
except Exception as e:
    print(e)
finally:
    atv.close()
```

but this gives me an issue with StreamReader not being seekable. 

I am planning to write a wrapper around the process so it meets the requirements of Pyatv.

I thought it might be wise to try and ask for help from an expert first though :) 

Love the library, happy to help contribute with Pull Requests if desired.

**Comment by daredoes** [comment]
Small update. My current work around is to stream the audio to a webpage in one script, and then stream that URL via pyatv instead. It works, except for significant delay from snapcast that I haven't yet figured out how to compensate for. 

Any advice is appreciated

```python
import asyncio
from aiohttp import web

async def create_process(cmd, *args):
    process = await asyncio.subprocess.create_subprocess_exec(
        cmd, stdin=None, stdout=asyncio.subprocess.PIPE, stderr=None, *args)
    return process

async def stream_pcm(request):
    resp = web.StreamResponse()
    resp.content_type = 'audio/mpeg'  # You might need to adjust the content type depending on your specific needs

    await resp.prepare(request)
    process = await create_process("./snapclient.sh")

    while True:
        data = await process.stdout.read(4096)
        if not data:
            break
        try:
          await resp.write(data)
        except ConnectionResetError:
            process.kill()

    return resp

app = web.Application()
app.router.add_get('/stream', stream_pcm)

web.run_app(app)
```

**Comment by postlund** [comment]
The reason why pyatv would seek in the input is to figure out metadata. Normally seeking isn't possible for streamed data and as long as the stream is marked as not being seekable, you should be fine. What does `process.stdout.seekable()` return in your first example (just before starting to stream)?

**Comment by daredoes** [comment]
So in this case, its that `seekable` isn't a function. Since the code inside RAOP does `if not self.reader.seekable():` without checking if seekable is available as a function. It's not present on `StreamReader`.

Thanks for the quick response @postlund!

**Comment by daredoes** [comment]
if I modify the function inside of RAOP to skip past this issue, the next issue I encounter is due to StreamReader being async 

```
* Starting to stream
('failed to init decoder', -1)
sys:1: RuntimeWarning: coroutine 'StreamReader.read' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
```

If I modify snapclient to pipe through ffmpeg to MP3 first, it causes a new issue
```
#!/bin/bash
snapclient -h snapcast.local --player file --logsink null --sampleformat 44100:16:* | ffmpeg -f s16le -ar 44100 -ac 2 -i pipe:0 -f mp3 pipe:1
```

issue is 
```
Guessed Channel Layout for Input Stream ù0.0 : stereo
Input ù0, s16le, from 'pipe:0':
  Duration: N/A, bitrate: 1411 kb/s
  Stream ù0:0: Audio: pcmès16le, 44100 Hz, stereo, s16, 1411 kb/s
Stream mapping:
  Stream ù0:0 -> ù0:0 (pcmès16le (native) -> mp3 (libmp3lame))
Output ù0, mp3, to 'pipe:1':
  Metadata:
    TSSE            : Lavf59.27.100
  Stream ù0:0: Audio: mp3, 44100 Hz, stereo, s16p
    Metadata:
      encoder         : Lavc59.37.100 libmp3lame
* Starting to stream=00:00:02.66 bitrate= 129.3kbits/s speed=1.67x    
('failed to init decoder', -1)86 bitrate= 128.3kbits/s speed=1.11x    
sys:1: RuntimeWarning: coroutine 'StreamReader.read' was never awaited
RuntimeWarning: Enable tracemalloc to get the object allocation traceback
avèinterleavedèwriteèframe(): Broken pipe                                                                                                                                    
    Last message repeated 4 times
Error writing trailer of pipe:1: Broken pipe
size=     174kB time=00:00:11.25 bitrate= 126.8kbits/s speed=1.12x    
video:0kB audio:176kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
Error closing file pipe:1: Broken pipe
Conversion failed!
```

**Comment by postlund** [comment]
I think the biggest issue is that `stream_file` does not support async based reader objects. Supporting that was a bit tricky since the reader is passed to miniaudio, which is a synchronous library. But I can probably give it a go and see if I can fix that. Would be great if you could provide me with a full example code and audio file so I can reproduce your particular usecase.

**Comment by daredoes** [comment]
Running on Mac OS X using `snapcast` installed via `brew`. Planning to move to a docker instance on alpine linux after development. 

Requirements.txt
```
aiofiles==23.1.0
aiohttp==3.8.4
aiosignal==1.3.1
async-timeout==4.0.2
attrs==23.1.0
bitarray==2.7.3
certifi==2023.5.7
cffi==1.15.1
chacha20poly1305-reuseable==0.2.5
charset-normalizer==3.1.0
cryptography==40.0.2
frozenlist==1.3.3
idna==3.4
ifaddr==0.2.0
mediafile==0.11.0
miniaudio==1.57
multidict==6.0.4
mutagen==1.46.0
protobuf==4.23.0
pyatv @ git+https://github.com/postlund/pyatv.git@e6d9299efa4f08fdb3a0f3ee89364bcdc10fec7d
pycparser==2.21
pydub==0.25.1
requests==2.30.0
six==1.16.0
srptools==1.0.1
urllib3==2.0.2
yarl==1.9.2
zeroconf==0.62.0
```

snapclient.sh
```bash
#!/bin/bash
snapclient -h YOUR_HOST_HERE --player file --logsink null 
```

test.py (this is cleaned up. If you want the messy tons of comments version let me know)
```python
"""Find devices on the network, and pick one to start streaming snapclient to

python test.py
"""

import asyncio
import asyncio.subprocess as asp
import sys


from typing import List
import pyatv
from pyatv.interface import Playing, PushListener
from pyatv.const import Protocol
from pyatv.conf import BaseConfig
from pyatv.helpers import is_streamable
from pyatv.protocols.raop.audio_source import ReaderWrapper

chunk_size = 1024
LOOP = asyncio.get_event_loop()

# Create my own wrapper for the audio buffer?
    

async def create_process(cmd, *args):
    process = await asp.create_subprocess_exec(
        cmd, stdin=None, stdout=asp.PIPE, stderr=None, *args)
    return process

def handle_device(device: BaseConfig):
    print(f"Found device: {device.name} {device.device_info.model}, address: {device.address}")

def get_valid_devices(devices: List[BaseConfig]):
    return devices



class PushUpdatePrinter(PushListener):
    """Print push updates to console."""

    def playstatus_update(self, updater, playstatus: Playing) -> None:
        """Inform about changes to what is currently playing."""
        print(30 * "-" + "\n", playstatus)

    def playstatus_error(self, updater, exception: Exception) -> None:
        """Inform about an error when updating play status."""
        print("Error:", exception)


async def pair(conf: BaseConfig, loop: asyncio.AbstractEventLoop):
    pairing = await pyatv.pair(conf, Protocol.RAOP, loop)
    await pairing.begin()

    pin = int(input("Enter PIN: "))
    pairing.pin(pin)
    await pairing.finish()

    # Give some feedback about the process
    if pairing.has_paired:
        print("Paired with device!")
        print("Credentials:", pairing.service.credentials)
    else:
        print("Did not pair with device!")

    await pairing.close()

async def stream_with_push_updates(
    conf: BaseConfig, loop: asyncio.AbstractEventLoop
):
    """Find a device and print what is playing."""
    print("* Connecting to", conf.address)
    
    # await pair(conf, loop)
    atv = await pyatv.connect(conf, loop)

    listener = PushUpdatePrinter()
    atv.push_updater.listener = listener
    atv.push_updater.start()

    process = await create_process("./snapclient.sh")
    # letting the script start up
    await asyncio.sleep(3)
    try:
        print("* Starting to stream")
        await atv.stream.stream_file(process.stdout)
        await asyncio.sleep(0.01)
    except Exception as e:
        print(e)
    finally:
        atv.close()

async def scan(
    loop: asyncio.AbstractEventLoop
):
    """Find a device and print what is playing."""
    print("* Discovering devices on network...")
    atvs = await pyatv.scan(loop, protocol=Protocol.RAOP)

    if not atvs:
        print("* Device found", file=sys.stderr)
        return
    devices = get_valid_devices(atvs)

    if not devices:
        print("* No Valid Devices found", file=sys.stderr)
        return
    await select_to_stream(devices, loop)
    return devices
async def select_to_stream(
    devices: List[BaseConfig],
    loop: asyncio.AbstractEventLoop
):
    """Find a device and print what is playing."""
    for i, device in enumerate(devices, start=1):
        print(f"{i}: {device.name} {device.device_info.model}, address: {device.address}")
    selection = -1
    while selection < 0:
        user_input = input("Pick a device by number: ")
        try:
            converted = int(user_input) - 1
            if len(devices) > converted:
                selection = converted
        except:
            print(f"{user_input} was not a valid device choice")
    
    device = devices[selection]

    await stream_with_push_updates(device, loop)



if __name__ == "__main__":
    LOOP.run_until_complete(scan(LOOP))
```

**Comment by postlund** [comment]
Thanks, I will look into it most likely on Wednesday or so.

**Comment by daredoes** [comment]
No rush! All the help is appreciated :)

**Comment by daredoes** [comment]
Bumping this so it doesn't fall of your radar! If I can get this to work effectively, it'll turn every HomePod in my house into a snapcast speaker!

**Comment by postlund** [comment]
It's a bit tricky to get everything right since miniaudio (the library I use for decoding) is synchronous, but the input to it now becomes async (via asyncio). But I have a concept in https://github.com/postlund/pyatv/tree/stream_asyncio, feel free to try that one out. I just made it quick and dirty, so it breaks all normal support (it only supports `asyncio.streams.StreamReader` as input to `stream_file`).

**Comment by daredoes** [comment]
This works! Now I just need to adjust for latency, but that's my issue, not pyatvs!

**Comment by postlund** [comment]
Great, thanks for trying! I will try to clean it up as soon as I can and merge it. Won't make any promises, but I'm making a release with #1975 tomorrow and maybe I could include this as well.

**Comment by daredoes** [comment]
@postlund before we consider this closed. I am only able to get it to work with an MP3 stream, and not a FLAC stream. Any advice on this?

**Comment by daredoes** [comment]
`Codec: flac, sampleformat: 44100:16:2`

**Comment by daredoes** [comment]
actually let me do more testing first. I let my computer sit on the FLAC stream for a bit, and some crazy sound popped in and scared me.

**Comment by daredoes** [comment]
I've got this working with the following command, still FLAC though.

```bash
snapclient -h snapcast.local --logsink null --player file  --sampleformat 44100:16:* --mixer script:./mixer_$hostID.sh --hostID $hostID $@ | ffmpeg -f u16le -acodec pcm_s16le -ac 2 -ar 44100 -i pipe:0 -f mp3 pipe:1 
```

The mixer script is as follows, it takes `--volume 0-100 --mute true|false` as args . It allows snapcast to control the airplay device's volume. Not sure how to get updates about the volume changes on the homepod yet though. Once I know that I can sync the changes back to the snapcast client.

```
#!/bin/bash

# Initialize parameters
volume=""
mute=""
hostID=""

# Parse command line parameters
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --volume) volume="$2"; shift ;;
        --mute) mute="$2"; shift ;;
        *) echo "Unknown parameter passed: $1"; exit 1 ;;
    esac
    shift
done

# Check if parameters are set
if [[ -z "$volume" || -z "$mute" ]]; then
    echo "Parameters --volume and --mute are required."
    exit 1
fi

# If mute is true, set volume to 0
if [[ "$mute" == "true" ]]; then
    volume=0
fi

volume=$(echo "scale=0; $volume * 100 / 1" | bc)

# Adjust this URL based on your requirements
curl "http://localhost:8080/volume/$hostID/$volume"
```

**Comment by postlund** [comment]
@michalmo is working on push updates for volume changes (it's described at the bottom [here](https://pyatv.dev/development/listeners/)). It will be part of the next release. If you want a specific volume, you can also set it prior to starting to stream. pyatv should set that volume instead of relying on what the speaker uses by default.

But I assume that this works good enough so I can make a proper change for it?

**Comment by daredoes** [comment]
It does work "good enough"!

**Comment by postlund** [comment]
Ok, that's good. Feel free to open additional issues for potential improvements later on. I'll try to get this fixed by tomorrow.

**Comment by postlund** [comment]
@daredoes I just opened #2010, can you verify if that still works as expected?
